nick_work/fdaaa_port.py

```diff
@@ -0,0 +1,372 @@
+            parsed_json = xmltodict.parse(xml_json.read())
+
+            #nct_id
+            nct_id = (soup.nct_id).text
+
+            #study_type
+            study_type = (soup.study_type).text
+
+            #has_certificate
+            hc = soup.disposition_first_submitted
+            if hc is not None:
+                has_certificate = 1
+            else:
+                has_certificate = 0
+
+            #phase
+            phase = (soup.phase).text    
+
+            #fda_reg_drug
+            if soup.is_fda_regulated_drug is None:
+                fda_reg_drug = None
+            else:
+                fda_reg_drug = (soup.is_fda_regulated_drug).text
+
+            #fda_reg_device
+            if soup.is_fda_regulated_device is None:
+                fda_reg_device = None
+            else:
+                fda_reg_device = (soup.is_fda_regulated_device).text
+
+            #primary_purpose
+            if soup.find('primary_purpose') is not None:
+                primary_purpose = soup.find('primary_purpose').text
+            else:
+                primary_purpose = None
+
+            #is_fda_regulated
+            df_row = fda_reg_df.loc[fda_reg_df['nct_id'] == nct_id]
+            if df_row.empty:
+                is_fda_regulated = None
+            else:
+                is_fda_regulated = df_row['is_fda_regulated'].values[0]
+
+            #study_status
+            study_status = (soup.overall_status).text
+
+            #start_date
+            sd = soup.start_date
+            if sd is not None:
+                try:
+                    start_date = datetime.strptime(sd.text, '%B %d, %Y').date()
+                except ValueError:
+                    start_date = datetime.strptime(sd.text, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
+            else:
+                start_date = None
+
+            #available_completion_date
+            pcd = soup.primary_completion_date
+            if pcd is not None:
+                try:
+                    primary_completion_date = datetime.strptime(pcd.text, '%B %d, %Y').date()
+                    defaulted_pcd = 0
+                except ValueError:
+                    primary_completion_date = datetime.strptime(pcd.text, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
+                    defaulted_pcd = 1
+            else:
+                primary_completion_date = None
+                defaulted_pcd = 0
+
+            cd = soup.completion_date
+            if cd is not None:
+                try:
+                    completion_date = datetime.strptime(cd.text, '%B %d, %Y').date()
+                    defaulted_cd = 0
+                except ValueError:
+                    completion_date = datetime.strptime(cd.text, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
+                    defaulted_cd = 1
+            else:
+                completion_date = None
+                defaulted_cd = 0
```

Comment:
The above three date operations involve a lot of copy-and-paste, which indicates another function should probably be used.

Should also include comments to explain the relativedelta choice.
Something like this:

```python
def string_to_date(datestr):
    """Convert text to date. If the day is missing, pick the last day of the current month.
    Returns (date, is_defaulted_date) tuple.
    """
    is_defaulted_date = False
    try:
        parsed_date = datetime.strptime(datestr, '%B %d, %Y').date()
    except ValueError:
        parsed_date = datetime.strptime(datestr, '%B %Y').date() + relativedelta(months=+1) - timedelta(days=1)
        is_defaulted_date = True
    return (is_defaulted_date, parsed_date)

is_defaulted_date, completion_date = string_to_date(soup.completion_date)
```

Reply:
done